include/ibamr/CellConvectiveOperator.h

```diff
@@ -0,0 +1,208 @@
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_CellConvectiveOperator
+#define included_IBAMR_CellConvectiveOperator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibamr/config.h>
+
+#include "ibamr/ConvectiveOperator.h"
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class CellConvectiveOperator is a abstract class for an implementation of
+ * a convective differencing operator.
+ */
+class CellConvectiveOperator : public ConvectiveOperator
+{
+public:
+    /*!
+     * \brief Class constructor.
+     */
+    CellConvectiveOperator(std::string object_name,
+                           SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_cell_var,
+                           int Q_min_ghost_cell_width,
+                           SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                           ConvectiveDifferencingType difference_form,
+                           std::vector<SAMRAI::solv::RobinBcCoefStrategy<NDIM>*> bc_coefs);
+
+    /*!
+     * \brief Default destructor.
+     */
+    ~CellConvectiveOperator() = default;
+
+    /*!
+     * \brief Interpolate a cell-centered field Q to a face-centered field q, possibly using the provided advection
+     * velocity field.
+     *
+     * \TODO Revisit this function name.
+     */
+    virtual void interpolateToFaceOnHierarchy(int q_interp_idx, int Q_cell_idx, int u_idx, bool synch_cf_bdry = true);
+
+    /*!
+     * \brief Evaluate the face-centered flux Q to a face-centered field q using the provided advection velocity field.
+     *
+     * \TODO Revisit this function name.
+     */
+    virtual void
+    evaluateAdvectiveFluxOnHierarchy(int q_flux_idx, int Q_cell_idx, int u_idx, const bool synch_cf_bdry = true);
+
+    /*!
+     * \brief Compute the advective derivative N = u * grad Q on the patch hierarchy.
+     *
+     * \TODO Revisit this function name.
+     */
+    virtual void
+    computeAdvectiveDerivativeOnHierarchy(int N_cell_idx, int q_interp_idx, int u_idx, bool synch_cf_bdry = true);
+
+    /*!
+     * \brief Compute the conservative derivative N = div(Q u) on the patch hierarchy.
+     *
+     * \TODO Revisit this function name.
+     */
+    virtual void computeConservativeDerivativeOnHierarchy(int N_cell_idx, int q_flux_idx, bool synch_cf_bdry = true);
+
+    /*!
+     * \brief Interpolate a cell-centered field Q to a face-centered field q on a single grid patch.
+     *
+     * \TODO Revisit this function name.
+     */
+    virtual void interpolateToFaceOnPatch(SAMRAI::pdat::FaceData<NDIM, double>& q_interp_data,
+                                          SAMRAI::pdat::CellData<NDIM, double>& Q_cell_data,
+                                          SAMRAI::pdat::FaceData<NDIM, double>& u_data,
```

Comment:
Should these be marked `const`?
```suggestion
                                          const SAMRAI::pdat::CellData<NDIM, double>& Q_cell_data,
                                          const SAMRAI::pdat::FaceData<NDIM, double>& u_data,
```